### ISSUE TITLE
helm cannot delete deployments that are in state ContainerCreating

### ISSUE THREAD

**Issue by lixiaobing1** [kubernetes/helm#2308]
use "kubectl delete deployment tiller-deploy --namespace kube-system, or more concisely helm reset"
but tiller service and pod still exist, so why?
Is there other method to delete or uninstall helm?

**Comment by bacongobbler** [comment]
I can't reproduce this on v2.3.1:

```
><> helm init
$HELM_HOME has been configured at /home/bacongobbler/.helm.

Tiller (the helm server side component) has been installed into your Kubernetes Cluster.
Happy Helming!
><> helm version
Client: &version.Version{SemVer:"v2.3.1", GitCommit:"32562a3040bb5ca690339b9840b6f60f8ce25da4", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.3.1", GitCommit:"32562a3040bb5ca690339b9840b6f60f8ce25da4", GitTreeState:"clean"}
><> helm reset
Tiller (the helm server side component) has been uninstalled from your Kubernetes Cluster.
><> kk get po
NAME                             READY     STATUS        RESTARTS   AGE
kube-addon-manager-minikube      1/1       Running       0          17m
kube-dns-v20-4b0ll               3/3       Running       0          17m
kubernetes-dashboard-xs3qx       1/1       Running       0          17m
tiller-deploy-1332173772-19hzx   0/1       Terminating   0          1m
```

The pod goes into Terminating status and is eventually removed from the cluster.

Can you provide more backstory/logs on how we can reproduce your issue? It's not clear to me what the issue is (there is no tiller service deployed on `helm init`).

**Comment by lixiaobing1** [comment]
my helm is abnormal, when I reset it, can't remove:

kubectl get pods --namespace kube-system
NAME                             READY     STATUS             RESTARTS   AGE
tiller-deploy-1332173772-jjnb8   0/1       ImagePullBackOff   0          54m

**Comment by bacongobbler** [comment]
I'm guessing this is related to #2309 then? Once that issue is resolved you should be able to run `helm reset`, though I'm curious if there's a reason we can't just make an API call to delete the deployment even if we can't connect to it.

**Comment by lixiaobing1** [comment]
now it can be deleted with kubectl cmd, but it is can't yesterday, thx very much:)

**Comment by lixiaobing1** [comment]
issue occurs again:

root@ubuntu:/home/zte# kubectl delete deployment tiller-deploy --namespace kube-system
deployment "tiller-deploy" deleted

root@ubuntu:/home/zte# kubectl get service --namespace kube-system
NAME            CLUSTER-IP   EXTERNAL-IP   PORT(S)     AGE
tiller-deploy   10.20.8.41   <none>        44134/TCP   21h

root@ubuntu:/home/zte# helm reset
Error: could not find a ready tiller pod

root@ubuntu:/home/zte# kubectl get service --namespace kube-system
NAME            CLUSTER-IP   EXTERNAL-IP   PORT(S)     AGE
tiller-deploy   10.20.8.41   <none>        44134/TCP   21h

**Comment by ReSearchITEng** [comment]
I would phrase it:
`helm delete` && `helm reset` do not delete a non-finished deployment (while IMHO it should).
I expect that `helm delete` && `helm reset` would do the equivalent of: 
`kubectl delete deployment --namespace=kube-system tiller-deploy`, but in the event of non-completed/stuck deployment it doesn't/
```
[root@machine0 ~]# kubectl get -o wide --namespace kube-system pods | tail -n +2
etcd-machine0                      1/1       Running             0          26m       10.234.176.71   machine0
kube-apiserver-machine0            1/1       Running             0          26m       10.234.176.71   machine0
kube-controller-manager-machine0   1/1       Running             0          26m       10.234.176.71   machine0
kube-dns-2245654080-nhq7l            3/3       Running             0          26m       10.244.0.10     machine0
kube-flannel-ds-27wzd                2/2       Running             0          26m       10.234.176.71   machine0
kube-flannel-ds-kjk3x                2/2       Running             0          25m       10.234.176.73   machine1
kube-flannel-ds-kkm60                2/2       Running             0          25m       10.234.176.77   machine2
kube-proxy-4ms64                     1/1       Running             0          25m       10.234.176.77   machine2
kube-proxy-h7152                     1/1       Running             0          25m       10.234.176.73   machine1
kube-proxy-tdlbr                     1/1       Running             0          26m       10.234.176.71   machine0
kube-scheduler-machine0            1/1       Running             0          26m       10.234.176.71   machine0
tiller-deploy-1332173772-rwrnw       0/1       ContainerCreating   0          4s        <none>          machine1
[root@machine0 ~]# helm reset
Error: could not find a ready tiller pod
[root@machine0 ~]# helm delete
Error: could not find a ready tiller pod
[root@machine0 ~]# kubectl get -o wide --namespace kube-system pods | tail -n +2
etcd-machine0                      1/1       Running             0          27m       10.234.176.71   machine0
kube-apiserver-machine0            1/1       Running             0          27m       10.234.176.71   machine0
kube-controller-manager-machine0   1/1       Running             0          27m       10.234.176.71   machine0
kube-dns-2245654080-nhq7l            3/3       Running             0          26m       10.244.0.10     machine0
kube-flannel-ds-27wzd                2/2       Running             0          26m       10.234.176.71   machine0
kube-flannel-ds-kjk3x                2/2       Running             0          25m       10.234.176.73   machine1
kube-flannel-ds-kkm60                2/2       Running             0          25m       10.234.176.77   machine2
kube-proxy-4ms64                     1/1       Running             0          25m       10.234.176.77   machine2
kube-proxy-h7152                     1/1       Running             0          25m       10.234.176.73   machine1
kube-proxy-tdlbr                     1/1       Running             0          26m       10.234.176.71   machine0
kube-scheduler-machine0            1/1       Running             0          26m       10.234.176.71   machine0
tiller-deploy-1332173772-rwrnw       0/1       ContainerCreating   0          20s       <none>          machine1
[root@machine0 ~]# kubectl delete deployment --namespace=kube-system tiller-deploy
deployment "tiller-deploy" deleted
[root@machine0 ~]# kubectl get -o wide --namespace kube-system pods | tail -n +2
etcd-machine0                      1/1       Running       0          27m       10.234.176.71   machine0
kube-apiserver-machine0            1/1       Running       0          27m       10.234.176.71   machine0
kube-controller-manager-machine0   1/1       Running       0          27m       10.234.176.71   machine0
kube-dns-2245654080-nhq7l            3/3       Running       0          26m       10.244.0.10     machine0
kube-flannel-ds-27wzd                2/2       Running       0          26m       10.234.176.71   machine0
kube-flannel-ds-kjk3x                2/2       Running       0          25m       10.234.176.73   machine1
kube-flannel-ds-kkm60                2/2       Running       0          25m       10.234.176.77   machine2
kube-proxy-4ms64                     1/1       Running       0          25m       10.234.176.77   machine2
kube-proxy-h7152                     1/1       Running       0          25m       10.234.176.73   machine1
kube-proxy-tdlbr                     1/1       Running       0          26m       10.234.176.71   machine0
kube-scheduler-machine0            1/1       Running       0          27m       10.234.176.71   machine0
tiller-deploy-1332173772-rwrnw       0/1       Terminating   0          33s       <none>          machine1
```

**Comment by bacongobbler** [comment]
Thank you for articulating what I was trying to point out, @ReSearchITEng. That's exactly what I was trying to say. :)

**Comment by lixiaobing1** [comment]
with "kubectl delete deployment", the pod is still terminating after several minutes.

**Comment by technosophos** [comment]
The deletion logic _should_ get fixed by 2.4 simply because we switched to the newer Kubernetes library.
